tests/pytests/functional/modules/test_saltutil.py

```diff
@@ -0,0 +1,33 @@
+import salt.config
+import salt.loader
+import salt.modules.saltutil
+import salt.state
+from tests.support.mock import patch
+
+
+@patch.object(salt.state.HighState, "destroy")
+@patch.object(salt.state.HighState, "top_matches")
+@patch.object(salt.state.HighState, "get_top")
+@patch.object(salt.state.State, "_gather_pillar")
+@patch("salt.utils.extmods.sync")
+@patch.object(salt.state.BaseHighState, "_BaseHighState__gen_opts")
+def test__get_top_file_envs(
+    gen_opts, sync, gather_pillar, get_top, top_matches, destroy, tmpdir
+):
+    """
+    Ensure we cleanup objects created by saltutil._get_top_file_envs #60449
+    """
+    opts = salt.config.minion_config("")
+    opts["master_uri"] = "tcp://127.0.0.1:11111"
+    opts["pki_dir"] = tmpdir
+    modules = salt.loader.minion_mods(opts, context={})
+    sync.return_value = (None, None)
+    # Mock the __gen_opts method of HighState so it doesn't try to auth to master.
+    gen_opts.return_value = opts
+    # Mock the _gather_pillar method of State so it doesn't try to auth to master.
+    gather_pillar.return_value = {}
+    top_matches.return_value = {}
+    modules["saltutil.sync_clouds"]()
+    assert get_top.called
+    # Ensure destroy is getting called
+    assert destroy.called
```

Comment:
```suggestion
import pytest
import salt.config
import salt.loader
import salt.modules.saltutil
import salt.state
from tests.support.mock import patch


@pytest.fixture
def opts(tmpdir):
    opts = salt.config.minion_config("")
    opts["master_uri"] = "tcp://127.0.0.1:11111"
    opts["pki_dir"] = str(tmpdir)
    yield opts


@pytest.fixture
def modules(opts):
    yield salt.loader.minion_mods(opts, context={})


@pytest.fixture
def configure_mocks(opts):
    with patch("salt.utils.extmods.sync", return_value=(None, None)):
        with patch.object(salt.state.HighState, "top_matches", return_value={}):
            # Mock the __gen_opts method of HighState so it doesn't try to auth to master.
            with patch.object(
                salt.state.BaseHighState, "_BaseHighState__gen_opts", return_value=opts
            ):
                # Mock the _gather_pillar method of State so it doesn't try to auth to master.
                with patch.object(salt.state.State, "_gather_pillar", return_value={}):
                    yield


@pytest.fixture
def destroy(configure_mocks):
    with patch.object(salt.state.HighState, "destroy") as destroy:
        yield destroy


@pytest.fixture
def get_top(configure_mocks):
    with patch.object(salt.state.HighState, "get_top") as get_top:
        yield get_top


def test__get_top_file_envs(modules, get_top, destroy):
    """
    Ensure we cleanup objects created by saltutil._get_top_file_envs #60449
    """
    modules["saltutil.sync_clouds"]()
    assert get_top.called
    # Ensure destroy is getting called
    assert destroy.called
```

This avoids [decorating the test function](https://engblog.nextdoor.com/what-your-mocks-do-when-you-aren-t-looking-b278e0d9e201) and also avoids mixing pytest fixture arguments injection with patch decorator arguments injection.
For pytest fixture arguments, the order doesn't matter since pytest will do the right thing, for patch decorator arguments, order matters and is prone to errors.

Reply:
I'm not sure I'm a fan of the multiple with statements piling up over decorators. Perhaps we can find a middle ground.

Reply:
Sure, as long as you don't decorate test functions :)

Turn each patch onto a fixture if it helps....

Seriously, using decorators is error prone(the order of decorating and the order of the injected arguments matters), specially when mixing with fixtures.
You have to know about that and now you're forcing anyone reading that test to know about that and that the order matters....

Also, your decorator patches, not all of them need to be passed to the test function.

Seriously, the patch decorators look like a hack to support injecting arguments into unittest tests. Pytest found a better way....

Reply:
If you could name the decorators and in then the function would be inspected to match the decorators with the arguments, no matter the order, then it would be a different ball game.

Reply:
Plus, there's that article which I believe still holds true today. Could be wrong here though.

Reply:
Yeah, this all makes sense to me.